Resolve `char` as a primitive even if there is a module in scope

jyn514 opened the pull request:
Closes https://github.com/rust-lang/rust/issues/58699.
r? @Manishearth

rust-highfive commented:
Some changes occurred in intra-doc-links.

cc @jyn514

Manishearth commented on src/librustdoc/passes/collect_intra_doc_links.rs:
```diff
@@ -217,15 +216,12 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
-                                    return Err(ErrorKind::AnchorFailure(AnchorFailure::Primitive));
-                                }
-                                return Ok((prim, Some(path_str.to_owned())));
+                        // This resolved to a module, but we want primitive types to take precedence instead.
```
We should avoid this branch if it's explicitly not specified as a type, yes?

jyn514 replied:
Oh you're right - this should happen if there's no disambiguator or if the disambiguator is `type@`, but if there's a different ambiguator we should skip it.

jyn514 replied:
Should be fixed now.

jyn514 committed:
Give precedence to primitives over modules

This has less surprising behavior when there is a module with the same
name as a primitive in scope.

Manishearth commented:
@bors r+

bors commented:
:pushpin: Commit 6d2b16fc7e9d82befdd864ec9327883424ac6668 has been approved by `Manishearth`

<!-- @bors r=Manishearth 6d2b16fc7e9d82befdd864ec9327883424ac6668 -->
<!-- homu: {"type":"Approved","sha":"6d2b16fc7e9d82befdd864ec9327883424ac6668","approver":"Manishearth"} -->

bors commented:
:hourglass: Testing commit 6d2b16fc7e9d82befdd864ec9327883424ac6668 with merge f78b2bee99acf60e0086aa75572697d05f35b3b7...
<!-- homu: {"type":"BuildStarted","head_sha":"6d2b16fc7e9d82befdd864ec9327883424ac6668","merge_sha":"f78b2bee99acf60e0086aa75572697d05f35b3b7"} -->

rust-log-analyzer commented:
The job `x86_64-gnu` of your PR [failed](https://github.com/rust-lang-ci/rust/runs/963043223) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://github.com/rust-lang-ci/rust/commit/f78b2bee99acf60e0086aa75572697d05f35b3b7/checks/963043223/logs), [raw log](https://github.com/rust-lang-ci/rust/commit/f78b2bee99acf60e0086aa75572697d05f35b3b7/checks/963043223/logs)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/src/librustdoc/passes/collect_intra_doc_links.rs at line 218:
                     }
                     Res::Def(DefKind::Mod, _) => {
                         // This resolved to a module, but we want primitive types to take precedence instead.
-                        if matches!(disambiguator, None | Some(Disambiguator::Namespace(Namespace::TypeNS))) {
+                        if matches!(
+                            disambiguator,
+                            None | Some(Disambiguator::Namespace(Namespace::TypeNS))
+                        ) {
                             if let Some(prim) = is_primitive(path_str, ns) {
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/src/librustdoc/passes/collect_intra_doc_links.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
                                 if extra_fragment.is_some() {
                                     return Err(ErrorKind::AnchorFailure(AnchorFailure::Primitive));
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test
== clock drift check ==
  local time: Sun Aug  9 05:24:22 UTC 2020
  network time: Sun, 09 Aug 2020 05:24:22 GMT
== end clock drift check ==
== end clock drift check ==
##[error]Process completed with exit code 1.
Terminate orphan process: pid (2871) (node)
Terminate orphan process: pid (2899) (python)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/963048745)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/963048745","builder_name":"checks-actions"} -->

jyn514 commented:
@bors r=manishearth

bors commented:
:pushpin: Commit b11e2f23411ec9aea4aa261c644f2c9e0df9457a has been approved by `manishearth`

<!-- @bors r=manishearth b11e2f23411ec9aea4aa261c644f2c9e0df9457a -->
<!-- homu: {"type":"Approved","sha":"b11e2f23411ec9aea4aa261c644f2c9e0df9457a","approver":"manishearth"} -->
